Reserved cluster state service

grcevski opened the pull request:
This PR adds the reserved cluster state controller, which part of the File based settings PR (https://github.com/elastic/elasticsearch/pull/86224).

This is a reusable component that has number of Reserved State Handler components that know how to update parts of the Elasticsearch state. This component is only used by the File Settings Service at the moment, but the intent is for modules and plugins to be able to call it to save settings and entities in the cluster state that cannot be changed by the end user. For example work on using this component to update state from plugins/modules, please see https://github.com/elastic/elasticsearch/issues/88341

Relates to https://github.com/elastic/elasticsearch/pull/86224.

Commit:
Immutable cluster state controller

Commit:
Merge master

Commit:
Refactor with smaller files

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

grcevski commented on server/src/main/java/org/elasticsearch/ExceptionsHelper.java:
I added this helper because I couldn't find a suitable one here that would return each throwable message. When we parse JSON, the messages of errors are reported on each throwable as it goes down in subsections of the JSON body.

rjernst commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateController.java:
```diff
@@ -0,0 +1,232 @@
+ * the cluster state specified in a file or through plugins/modules. Reserved cluster state
+ * cannot be modified throught the REST APIs, only through this controller class.
+ */
+public class ReservedClusterStateController {
```
nit: we've typically used the "Service" suffix for other similar "things" in a node, any reason not to use that here in place of Controller?

rjernst commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java:
```diff
@@ -0,0 +1,138 @@
+            try {
+                Set<String> existingKeys = keysForHandler(existingMetadata, handlerName);
+                TransformState transformState = handler.transform(reservedState.get(handlerName), new TransformState(state, existingKeys));
+                state = transformState.state();
```
Each handler is create it's own ClusterState. Can we do this more efficiently? Building cluster state is not cheap. Could we instead create a ClusterState.Builder here (or perhaps just the customs and metadata builders), and that is passed to the transform method? I think the TransformState would not be needed, the method could just return the new set of keys.

grcevski replied:
I can't fully use ClusterState.Builder without more significant refactor of how the TransportActions we are wrapping for this work. All of the transport actions we use for the reserved state handlers take cluster state in and return a new state back. However, I really like the idea of using the builder instead of the TransportState. This way we can remove that TransportState class, we can start using the builder and then we can slowly improve the transport actions to maybe work with a builder in the future, to improve efficiency.

grcevski replied:
I think I said that too quickly, I think using the builder at the moment without refactoring all of the transport actions will be more overhead. I'll rip out transport state though, I can simply use Tuple.

rjernst replied:
I'm ok with waiting on this optimization, but I do think it is something we should look into. Creating cluster state is relatively expensive.

grcevski replied:
I'll follow-up with a PR to optimize this after I check that it's OK to refactor the Transport node actions to use a builder.

rjernst commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java:
```diff
@@ -0,0 +1,138 @@
+
+                throw new ReservedStateVersion.IncompatibleVersionException(
+                    format(
+                        "Not updating error state because version [%s] is less or equal to the last state error version [%s]",
```
Why is this an error? Isn't this part of the idempotent nature of the reserved state design?

rjernst replied:
What about if the same operator settings file has an update triggered, but for the same version? Wouldn't this cause an internal error, when it should be a no-op?

Separate from this PR, I think we should consider how plugins can use this functionality without directly calling the reserved state service. As you point out, it can be error prone, so perhaps we can redesign the pluggable portion to declaratively specify their resources, and the server is actually what does the registration.

grcevski replied:
Fair point, I throw a specific exception here that would be caught and handled differently by the caller of the service, but it's not ideal, it should be a no-op and a log message only. I'll fix this.

grcevski replied:
> Separate from this PR, I think we should consider how plugins can use this functionality without directly calling the reserved state service. As you point out, it can be error prone, so perhaps we can redesign the pluggable portion to declaratively specify their resources, and the server is actually what does the registration.

I like this idea, perhaps though SPI we can ask all plugins if they want to declare any state chunks that need to be applied and we call the update service internally. This way no plugins will have access to the update service and we'll apply it using our rules.

rjernst commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateVersion.java:
```diff
@@ -0,0 +1,58 @@
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), COMPATIBILITY);
+    }
+
+    public static ReservedStateVersion parse(XContentParser parser, Void v) {
```
Could the implementation specific Void here could be hidden from the signature? Remove from the method and pass null to PARSER?

rjernst reviewed (commented):
Looks pretty good. I left some thoughts.

grcevski commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateErrorTaskExecutor.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void clusterStatePublished(ClusterState newClusterState) {
+        logger.info("Wrote new error state in immutable metadata");
```
I think it's useful to know that we've recorded the error state in the cluster state. Updating the state can fail here, maybe we are not the master anymore, or ES terminated before we got to write the error state. By seeing the message in the logs, we definitely know we wrote what's wrong with the file settings. Thanks for catching the wording bug.

grcevski replied:
Yeah that's a valid point, it's not useful to the end user as is the confirmation of new cluster state being set. I'll change this to a debug message instead, so we can use it while working on the feature.

Commit:
Apply review feedback

Commit:
Apply more review feedback

grcevski commented:
Thanks again for the review @rjernst, I removed that extra exception and changed such that we return the original cluster state reference to avoid unnecessary update.

rjernst commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java:
```diff
@@ -179,31 +177,23 @@ public void onFailure(Exception e) {
-                    )
+            logger.warn(
+                () -> format(
+                    "Cluster state version [%s] is not compatible with this Elasticsearch node",
```
Perhaps this should mention the reserved state name?

rjernst reviewed (approved):
LGTM

Commit:
Add namespace to log messages

grcevski commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into operator/just_controller
